src/engine/Server.cpp

```diff
@@ -882,24 +882,47 @@ Awaitable<void> Server::processUpdate(
                      "following query was sent instead of an update: ",
                      plannedQuery.parsedQuery_._originalString));
   }
-
-  // Update the delta triples.
-  //
-  // TODO: This does not yet handle concurrent updates correctly, but it should
-  // work fine when a new update is sent only after the previous one has
-  // finished.
-  auto& deltaTriplesManager = index_.deltaTriplesManager();
-  deltaTriplesManager.modify(
-      [this, &plannedQuery, &qet, &cancellationHandle](auto& deltaTriples) {
-        ExecuteUpdate::executeUpdate(index_, plannedQuery.parsedQuery_, qet,
-                                     deltaTriples, cancellationHandle);
-      });
+  ExecuteUpdate::executeUpdate(index_, plannedQuery.parsedQuery_, qet,
+                               deltaTriples, cancellationHandle);
 
   LOG(INFO) << "Done processing update"
             << ", total time was " << requestTimer.msecs().count() << " ms"
             << std::endl;
   LOG(DEBUG) << "Runtime Info:\n"
              << qet.getRootOperation()->runtimeInfo().toString() << std::endl;
+}
+
+// ____________________________________________________________________________
+Awaitable<void> Server::processUpdate(
+    const ad_utility::url_parser::ParamValueMap& params, const string& update,
+    ad_utility::Timer& requestTimer,
+    const ad_utility::httpUtils::HttpRequest auto& request, auto&& send,
+    TimeLimit timeLimit) {
+  auto messageSender = createMessageSender(queryHub_, request, update);
+
+  auto [cancellationHandle, cancelTimeoutOnDestruction] =
+      setupCancellationHandle(messageSender.getQueryId(), timeLimit);
+
+  // Update the delta triples.
+  co_await computeInNewThread(
+      updateThreadPool_,
+      [this, &params, &update, &requestTimer, &timeLimit, &messageSender,
+       &cancellationHandle] {
+        index_.deltaTriplesManager().modify(
+            [this, &params, &update, &requestTimer, &timeLimit, &messageSender,
+             &cancellationHandle](auto& deltaTriples) {
```

Comment:
There are these false-positive warnings for the `this ` capture on clang again. You can mitigate them by explicitly and redundantly using `this->` at least once in each of the lambdas.